### PR TITLE
chore: support nix shell dev environment

### PR DESCRIPTION
**Pull request by jetjinser**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Provides a [nix shell](https://wiki.nixos.org/wiki/Flakes#Super_fast_nix-shell) for nix users.

### CODE REVIEW

File: docs/developer-guide.md
Comment (stdrc) [inline docs/developer-guide.md:111]: Please avoid use single `\n` to break line, because the rendered page won't break.

File: docs/developer-guide.md
Comment (stdrc) [inline docs/developer-guide.md:117]: Nit but there's an unnecessary new line here.

File: docs/developer-guide.md
Comment (stdrc) [inline docs/developer-guide.md:87]: May add a new line below markdown titles.

**Review by stdrc — COMMENTED** [review]
I'm afraid the `.envrc` and `nix` are too intrusive. AFAIK, few people are developing RW with nix shell so that the new file and folder introduced can be out of maintenance very quickly. Is it possible to move them to standalone repo or just put it in the developer guide and fold the section?

**Comment by jetjinser** [comment]
> I'm afraid the `.envrc` and `nix` are to intrusive. AFAIK, few people are developing RW with nix shell so that the new file and folder introduced can be out of maintenance very quickly. Is it possible to move them to standalone repo or just put it in the developer guide and fold the section?

I still hope it can be in the project.
1. I think people who don't need nix should simply ignore it.
2. The config shouldn't require much maintenance and we can mark it as non-first class support. Nix users should be able to figure it out themselves when it's broken.

besides,
1. It is possible to put it in a separate repo, but it will become troublesome to update it
2. Developers manually configure the nix shell from the developer guide will also cause trouble: this trouble is caused by the mechanism of nix flake, it will ignore files that are not tracked by git, which means developers need to be careful to manually keep nix/flake.nix tracked.

**Review by xxchan — COMMENTED** [review]
Generally same feeling as RC (btw previously I rejected similar one in `sqllogictest-rs` https://github.com/risinglightdb/sqllogictest-rs/pull/156

There are indeed some rationale not to add things to the repo if it's not useful for ALL developers (btw I also rejected 1 PR because of this reason https://github.com/risingwavelabs/risingwave/pull/15483#discussion_r1514115655): 
- RC already mentioned the maintenance issue.
- There's also confusion issue: For those not familiar with `nix`, they don't understand what it is, and might **misunderstood it's required** for development. 

Another problem is too many things in the root directory: https://github.com/risingwavelabs/risingwave/issues/17186.

---

But I'm also interested 
- Without these, how will `nix` users (e.g., you) develop? Are there any inconveniences? In other words, how much can they be benefited? 
- Are there any large projects having these? I've never seen one before 🤔

Welcome to convince me!

**Comment by jetjinser** [comment]
thank you for the explanation!

1. In fact, there are two situations here.
    - For users _using nix in Darwin or Linux distributions_ to develop RW, nix shell is not necessary because they can benefit from [FHS](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/index.html) (the corresponding dynamic link library can be found when building)
    - For _[NixOS](https://nixos.org/)_ users developing RW, nix shell becomes **necessary**. Although nix can install software and dependencies globally, environment variables such as `LD_LIBRARY_PATH`, `PKG_CONFIG_PATH`, etc. that help ld (etc.) find dependencies are generally set through the nix shell. The nix shell is per-project, and more nix users also prefer to organize dependencies individually in each project rather than installing them globally.
    BTW, I need to switch to the branch that have nix shell, enter the nix shell and then switch back to develop when I develop RW.

    In either case, nix users can benefit from the nix shell provided by the project. Contributing developers do **not need to manually install dependencies or worry about versions**, they just type `nix develop`, wait a moment, and everything is set.

2. Yes, there are some projects (developed with nix help).
    - https://github.com/element-hq/synapse
    - https://github.com/sile-typesetter/sile
    - https://github.com/aurae-runtime/aurae
    - https://github.com/eza-community/eza
    - https://github.com/hyprwm/Hyprland

**Comment by BugenZhao** [comment]
> I'm afraid the `.envrc` and `nix` are too intrusive. AFAIK, few people are developing RW with nix shell so that the new file and folder introduced can be out of maintenance very quickly. Is it possible to move them to standalone repo or just put it in the developer guide and fold the section?

However, I feel like the nix users generally have strong hands-on skills, like voluntarily contributing nix support to these open-source projects. 😄 If so, then maintenance won't be a big problem.

File: .envrc
Comment (xxchan) [inline .envrc:1]: Is this necessary? I don't find it in `synapse`, and `.envrc` sounds like very general, and thus a little more mysterious.

File: nix/devshell.nix
Comment (xxchan) [inline nix/devshell.nix:1]: I'd love to see comments (or a README) explaining what these files are, and how to use them.

https://github.com/element-hq/synapse/blob/7c169f4d2ccfa7e06c6f50a08475eadd1a9202dd/flake.nix#L1-L38

**Review by xxchan — COMMENTED** [review]
Thanks for the detailed explanations. It makes some sense to me. LGTM after comments resolved. But I want to have another person to click the green button.

(Btw, the "intrusive" nature of NixOS is still a little confusing to me. I'm not sure it's good or not)

File: .envrc
Comment (jetjinser) [reply .envrc:1]: No, this is _not necessary_. This is used by [direnv](https://github.com/direnv/direnv), a tool widely used by nix users who use it to automatically activate nix shell environments when enter the project dir.
This isn't very critical, just a handy tool.

**Comment by stdrc** [comment]
Is the `nix` must lie in root directory? If possible, I think we can put in somewhere like `develop/nix`, `tools/nix` or `dev_tools/nix`.

File: .envrc
Comment (stdrc) [reply .envrc:1]: If `.envrc` can be gitignored, I think we don't need to put it in main branch.

**Comment by jetjinser** [comment]
> Is the `nix` must lie in root directory? If possible, I think we can put in somewhere like `develop/nix`, `tools/nix` or `dev_tools/nix`.

As long as it is tracked by nix, it doesn't matter where it is.
But I guess this can wait for a future reorganization #17186? Perhaps it would be easier to move `nix/` to a more appropriate location when considering the entire project directory.

File: .envrc
Comment (jetjinser) [reply .envrc:1]: > If `.envrc` can be gitignored, I think we don't need to put it in main branch.

Let me try one more time :p

If we don't include it in the repository, `nix` + `direnv` users will have to keep an _untracked_ .`envrc` file in their project and be careful not to accidentally `git add` it.

But as mentioned before, this isn't very critical; it would just be `nice to have` 😄

**Comment by stdrc** [comment]
> But I guess this can wait for a future reorganization #17186?

No dependency I guess. The point here is to avoid introduce too-specific root-level folders.

File: .envrc
Comment (xxchan) [reply .envrc:1]: > be careful not to accidentally git add it.

FYI you can have local ignore file in `.git/info/exclude`

Comment (jetjinser) [reply .envrc:1]: > FYI you can have local ignore file in `.git/info/exclude`

Thank you for telling me. I should have thought of that earlier!

Comment (jetjinser) [reply .envrc:1]: Removed `.envrc`, added comments about `.git/info/exclude`.

File: develop/nix/devshell.nix
Comment (stdrc) [inline develop/nix/devshell.nix:20]: Is it possible to use `./risedev install-tools`? So that we can reduce the places we need to maintain the test/build/check dependencies.

File: develop/nix/devshell.nix
Comment (stdrc) [inline develop/nix/devshell.nix:42]: Better to add some comment here to remind future maintainers that these pkgs are listed in `developer-guide.md`'s "Set up the development environment" subsection. Always point to the source of trust.

File: develop/nix/flake.lock
Comment (stdrc) [inline develop/nix/flake.lock:1]: Does this file imply that developers using Nix will install fixed versions of 
dev dependencies? Is there a way to auto bump these deps? For example by utilizing dependabot? Not a strong opinion though.

File: develop/nix/devshell.nix
Comment (jetjinser) [reply develop/nix/devshell.nix:20]: Since `cargo install` compiles from source, yes, it can be used by nix users too.

But I think we can still keep it here? These tools are no different from other dependencies, we don't need much maintenance, nix users should figure it out by themself.

Comment (jetjinser) [reply develop/nix/devshell.nix:20]: Since `cargo install` compiles from source, yes, nix users can use it. I'll delete these if you think it's better.

File: develop/nix/flake.lock
Comment (jetjinser) [reply develop/nix/flake.lock:1]: I guess automatically bumping version is not a good idea.

Nix pursues a reproducible build. No matter which machine it is on, it is the same environment whenever you enter the nix shell. I think this may be more important for the development environment :)

**Comment by stdrc** [comment]
@xxchan What do you think? The current changes look actually pretty clean to me. The only issue remains future maintenance. Personally I tend to approve this PR.

**Comment by xxchan** [comment]
@stdrc As I said before, +1 after another +1, because I'm open to it, but not very enthusiastic :p

@neverchanje I understand your concern, but it seems the problems (why I don't like it too much, but it might still be needed and acceptable for some developers) are already discussed above.

> We welcome contributions that are feature-oriented and beneficial to all RisingWave users. For developers, let's keep things simple.

I don't think this is a convincing reason to reject this PR and should be a guideline for us. Developer experience actually matters a lot to me. 

"DX is a multiplier for basically everything. Whether you want to ship more features, to make code faster, to adapt to a change of requirements, or to attract new contributors, DX is a factor in that."

And I also favor simplicity a lot. That's what I raised at the beginning. 

I think we should be specific and constructive on the topic.

**Comment by xxchan** [comment]
> As I said before, +1 after another +1, because I'm open to it, but not very enthusiastic :p

https://www.apache.org/foundation/voting.html#expressing-votes-1-0-1-and-fractions

+0.5 😁

File: docs/developer-guide.md
Comment (stdrc) [inline docs/developer-guide.md:42]: Is this intended?

**Review by stdrc — APPROVED** [review]
Tried this PR in `nixos/nix` docker container, very smooth experience. LGTM

**Pull request merged**
